### ISSUE TITLE
Incorrect HAML string interpolation highlighting

### ISSUE THREAD

**Issue by LukeeeeBennett** [jneen/rouge#542]
I experienced an issue with incorrect HAML highlighting on [this gitlab diff line](https://gitlab.com/gitlab-org/gitlab-ce/blob/9541828e17350bda368ab643e97534b3f5904461/app/views/projects/merge_requests/widget/_heading.html.haml#L8).

![http://i.imgur.com/uXzl9Y6.png](http://i.imgur.com/uXzl9Y6.png)

After playing abou on http://rouge.jneen.net/, I found that the following breaks the highlighting.

``` haml
.some_class{ class: "other_#{class}" }
  = some_method "##{hash.prop}"
```

but the following un-breaks the highlighting.

``` haml
.some_class{ class: "other_class" }
  = some_method "##{hash.prop}"
```

More interestingly, we don't need the `#{` on the first interpolation to break the second, it seems to be the existence of `}` in the string before the interpolation, for example, the following still breaks the highlighting.

``` haml
.some_class{ class: "other_}class" }
  = some_method "##{hash.prop}"
```

![http://i.imgur.com/vNFgq8N.png](http://i.imgur.com/vNFgq8N.png)

**Comment by brodock** [comment]
It looks to be this rule:

```ruby
state :tag do
  #...
  rule(/\{#{comma_dot}*?\}/) { delegate ruby }
  #...
end
```

https://github.com/jneen/rouge/blob/d989f18440335b5098c6955a219b9e6abed8da35/lib/rouge/lexers/haml.rb#L134

I'm not 100% sure how to code that, but should be something like:

it should start a new state with `{` delegate it to ruby and push a check for '}'... while in ruby state, and when it's a string after the last unescaped ", pop!.

**Comment by brodock** [comment]
With the provided use-case changing the rule to:

```
rule(/\{#{comma_dot}*?\}$/) { delegate ruby }
```

worked, but I'm unsure yet if it's acceptable or not.

**Comment by jneen** [comment]
Oh... this is bad...

So it looks like we can't tell where the end of an interpolation is without parsing the ruby code. Ruby is... a really complicated lexer. We're basically going to have to do a much fancier delegation scheme that... knows about where ruby strings start and end... which is really complicated...

**Comment by jneen** [comment]
We could do something a bit half-assed to fix 90% of cases though:

``` ruby
rule(/\{(".*?"|'.*?'|#{comma_dot})*?)\}/) { ... }
```

But that doesn't catch triple-quoted strings, %-strings, or heredocs, just to start

**Comment by jneen** [comment]
Or... hang on lemme see if I can come up with something...

**Comment by jneen** [comment]
So, to start, ff9bbf6 contains the damage by resetting the ruby lexer at the beginning of each ruby block (so that it doesn't start in the string state as in your example)

**Comment by jneen** [comment]
Ooookay so I've made a discovery about how HAML itself processes these - it appears to only look for the balanced brace, disregarding string states. So the example you gave is actually an error:

``` shell
; echo '.foo{name: "bar}"}' | haml -s
Syntax error on line 131: (haml):1: syntax error, unexpected ',', expecting =>
..., name: "bar)}>\"}</div>\n", 0, false);;_erbout
...                               ^
  Use --trace for backtrace.
```

I think as long as we can replicate the balancing behavior we'll be fine.

**Comment by jneen** [comment]
Okay I believe I've fixed this in a294eeb, could you confirm?

**Comment by jneen** [comment]
Closing as fixed in v2.0.7. Thanks for the report!

**Comment by LukeeeeBennett** [comment]
I seem to have missed all this conversation but this is **aweeeesome**, thank you!
